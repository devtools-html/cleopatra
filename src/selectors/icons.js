/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';
import type { IconWithClassName, IconState } from '../types/state';
import type { Selector, DangerousSelectorWithArguments } from '../types/store';
import type { CallNodeDisplayData } from '../types/profile-derived';

/**
 * A simple selector into the icon state.
 */
export const getIcons: Selector<IconState> = state => state.icons;

/**
 * In order to load icons without multiple requests, icons are created through
 * CSS. This function gets the CSS class name for a call node. This function
 * does not perform any memoization, and updates every time. It could be updated
 * to memoize.
 */
export const getIconClassNameForCallNode: DangerousSelectorWithArguments<
  string,
  CallNodeDisplayData
> = (state, displayData) => {
  const icons = getIcons(state);
  return displayData.icon !== null && icons.has(displayData.icon)
    ? _classNameFromUrl(displayData.icon)
    : '';
};

/**
 * This functions returns an object with both the icon URL and the class name.
 */
export const getIconsWithClassNames: Selector<
  IconWithClassName[]
> = createSelector(
  getIcons,
  icons =>
    [...icons].map(icon => ({ icon, className: _classNameFromUrl(icon) }))
);

/**
 * Transforms a URL into a valid CSS class name.
 */
function _classNameFromUrl(url): string {
  return url.replace(/[/:.+>< ~()#,]/g, '_');
}
