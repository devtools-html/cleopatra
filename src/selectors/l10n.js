/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type { Selector, L10nState, Localization } from 'firefox-profiler/types';

export const getL10nState: Selector<L10nState> = state => state.l10n;
export const getIsL10nFetching: Selector<boolean> = state =>
  getL10nState(state).isL10nFetching;
export const getLocalization: Selector<Localization> = state =>
  getL10nState(state).localization;
