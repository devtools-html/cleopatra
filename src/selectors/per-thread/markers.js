/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';
import * as UrlState from '../url-state';
import * as MarkerData from '../../profile-logic/marker-data';
import * as MarkerTiming from '../../profile-logic/marker-timing';
import * as ProfileSelectors from '../profile';

import type {
  SamplesTable,
  MarkersTable,
  IndexIntoMarkersTable,
} from '../../types/profile';
import type {
  TracingMarker,
  MarkerTimingRows,
} from '../../types/profile-derived';
import type { Selector } from '../../types/store';
import type { $ReturnType } from '../../types/utils';
import type { Milliseconds } from '../../types/units';

/**
 * Infer the return type from the getMarkerSelectorsPerThread function. This
 * is done that so that the local type definition with `Selector<T>` is the canonical
 * definition for the type of the selector.
 */
export type MarkerSelectorsPerThread = $ReturnType<
  typeof getMarkerSelectorsPerThread
>;

/**
 * Create the selectors for a thread that have to do with either markers.
 */
export function getMarkerSelectorsPerThread(threadSelectors: *) {
  const _getMarkersTable: Selector<MarkersTable> = state =>
    threadSelectors.getThread(state).markers;

  const _getRangeFilteredThreadSamples: Selector<SamplesTable> = createSelector(
    threadSelectors.getRangeFilteredThread,
    thread => thread.samples
  );

  const getJankInstances: Selector<TracingMarker[]> = createSelector(
    _getRangeFilteredThreadSamples,
    (samples): TracingMarker[] => MarkerData.getJankInstances(samples, 50)
  );

  /**
   * Similar to thread filtering, the markers can be filtered as well, and it's
   * important to use the right type of filtering for the view. The steps for filtering
   * markers are a bit different, since markers can be valid over ranges, and need a
   * bit more processing in order to get into a correct state. There are a few
   * variants of the selectors that are created for specific views that have been
   * omitted, but the ordered steps below give the general picture.
   *
   * 1. _getMarkersTable - Get the MarkersTable from the current thread.
   * 2. getProcessedMarkersTable - Process marker payloads out of raw strings, and other
   *                               future processing needs. This returns a MarkersTable
   *                               still.
   * 3. getTracingMarkers - Match up start/end markers, and start returning
   *                        TracingMarkers.
   * 4. getCommittedRangeFilteredTracingMarkers - Apply the commited range.
   * 5. getSearchFilteredTracingMarkers - Apply the search string
   * 6. getPreviewFilteredTracingMarkers - Apply the preview range
   */
  const getProcessedMarkersTable: Selector<MarkersTable> = createSelector(
    _getMarkersTable,
    threadSelectors.getStringTable,
    MarkerData.extractMarkerDataFromName
  );

  const _getFirstSampleTime: Selector<Milliseconds> = state =>
    threadSelectors.getThread(state).samples.time[0] || 0;

  const getTracingMarkers: Selector<TracingMarker[]> = createSelector(
    getProcessedMarkersTable,
    threadSelectors.getStringTable,
    _getFirstSampleTime,
    MarkerData.getTracingMarkers
  );

  const getCommittedRangeFilteredTracingMarkers: Selector<
    TracingMarker[]
  > = createSelector(
    getTracingMarkers,
    ProfileSelectors.getCommittedRange,
    (markers, range): TracingMarker[] => {
      const { start, end } = range;
      return MarkerData.filterTracingMarkersToRange(markers, start, end);
    }
  );

  const getCommittedRangeFilteredTracingMarkersForHeader: Selector<
    TracingMarker[]
  > = createSelector(
    getCommittedRangeFilteredTracingMarkers,
    (markers): TracingMarker[] =>
      markers.filter(
        tm =>
          tm.name !== 'GCMajor' &&
          tm.name !== 'BHR-detected hang' &&
          tm.name !== 'LongTask' &&
          tm.name !== 'LongIdleTask' &&
          !MarkerData.isNetworkMarker(tm)
      )
  );

  const getSearchFilteredTracingMarkers: Selector<
    TracingMarker[]
  > = createSelector(
    getCommittedRangeFilteredTracingMarkers,
    UrlState.getMarkersSearchString,
    MarkerData.getSearchFilteredTracingMarkers
  );

  const getPreviewFilteredTracingMarkers: Selector<
    TracingMarker[]
  > = createSelector(
    getSearchFilteredTracingMarkers,
    ProfileSelectors.getPreviewSelection,
    (markers, previewSelection) => {
      if (!previewSelection.hasSelection) {
        return markers;
      }
      const { selectionStart, selectionEnd } = previewSelection;
      return MarkerData.filterTracingMarkersToRange(
        markers,
        selectionStart,
        selectionEnd
      );
    }
  );

  const getIsNetworkChartEmptyInFullRange: Selector<boolean> = createSelector(
    getTracingMarkers,
    markers => markers.filter(MarkerData.isNetworkMarker).length === 0
  );

  const getNetworkChartTracingMarkers: Selector<
    TracingMarker[]
  > = createSelector(getSearchFilteredTracingMarkers, markers =>
    markers.filter(MarkerData.isNetworkMarker)
  );

  const getMergedNetworkChartTracingMarkers: Selector<
    TracingMarker[]
  > = createSelector(
    getNetworkChartTracingMarkers,
    MarkerData.mergeStartAndEndNetworkMarker
  );

  const getIsMarkerChartEmptyInFullRange: Selector<boolean> = createSelector(
    getTracingMarkers,
    markers => MarkerData.filterForMarkerChart(markers).length === 0
  );

  const getMarkerChartTracingMarkers: Selector<
    TracingMarker[]
  > = createSelector(
    getSearchFilteredTracingMarkers,
    MarkerData.filterForMarkerChart
  );

  const getMarkerChartTiming: Selector<MarkerTimingRows> = createSelector(
    getMarkerChartTracingMarkers,
    MarkerTiming.getMarkerTiming
  );

  const getNetworkChartTiming: Selector<MarkerTimingRows> = createSelector(
    getNetworkChartTracingMarkers,
    MarkerTiming.getMarkerTiming
  );

  const getNetworkTracingMarkers: Selector<TracingMarker[]> = createSelector(
    getCommittedRangeFilteredTracingMarkers,
    tracingMarkers => tracingMarkers.filter(MarkerData.isNetworkMarker)
  );

  const getNetworkTrackTiming: Selector<MarkerTimingRows> = createSelector(
    getNetworkTracingMarkers,
    MarkerTiming.getMarkerTiming
  );

  const getScreenshotsById = createSelector(
    _getMarkersTable,
    threadSelectors.getStringTable,
    ProfileSelectors.getProfileRootRange,
    MarkerData.extractScreenshotsById
  );

  const getRangeFilteredScreenshotsById: Selector<
    Map<string, TracingMarker[]>
  > = createSelector(
    getScreenshotsById,
    ProfileSelectors.getCommittedRange,
    (screenshotsById, { start, end }) => {
      const newMap = new Map();
      for (const [id, screenshots] of screenshotsById) {
        newMap.set(
          id,
          MarkerData.filterTracingMarkersToRange(screenshots, start, end)
        );
      }
      return newMap;
    }
  );

  const getSelectedMarkerIndex: Selector<IndexIntoMarkersTable | -1> = state =>
    threadSelectors.getViewOptions(state).selectedMarker;

  return {
    getJankInstances,
    getProcessedMarkersTable,
    getTracingMarkers,
    getNetworkChartTracingMarkers,
    getIsMarkerChartEmptyInFullRange,
    getMarkerChartTracingMarkers,
    getMarkerChartTiming,
    getNetworkChartTiming,
    getCommittedRangeFilteredTracingMarkers,
    getCommittedRangeFilteredTracingMarkersForHeader,
    getNetworkTracingMarkers,
    getNetworkTrackTiming,
    getMergedNetworkChartTracingMarkers,
    getRangeFilteredScreenshotsById,
    getSearchFilteredTracingMarkers,
    getPreviewFilteredTracingMarkers,
    getSelectedMarkerIndex,
    getIsNetworkChartEmptyInFullRange,
  };
}
