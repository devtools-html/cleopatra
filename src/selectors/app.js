/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import { getSelectedTab } from './url-state';
import { tabSlugs } from '../app-logic/tabs-handling';
import { selectedThreadSelectors } from './per-thread';

import type { TabSlug } from '../app-logic/tabs-handling';
import type { AppState, AppViewState } from '../types/state';
import type { Selector } from '../types/store';

/**
 * Simple selectors into the app state.
 */
export const getApp: Selector<AppState> = state => state.app;
export const getView: Selector<AppViewState> = state => getApp(state).view;
export const getIsUrlSetupDone: Selector<boolean> = state =>
  getApp(state).isUrlSetupDone;
export const getHasZoomedViaMousewheel: Selector<boolean> = state => {
  return getApp(state).hasZoomedViaMousewheel;
};
export const getIsSidebarOpen: Selector<boolean> = state =>
  getApp(state).isSidebarOpenPerPanel[getSelectedTab(state)];
export const getPanelLayoutGeneration: Selector<number> = state =>
  getApp(state).panelLayoutGeneration;
export const getLastVisibleThreadTabSlug: Selector<TabSlug> = state =>
  getApp(state).lastVisibleThreadTabSlug;

/**
 * Visible tabs are computed based on the current state of the profile. Some
 * effort is made to not show a tab when there is no data available for it.
 */
export const getVisibleTabs: Selector<$ReadOnlyArray<TabSlug>> = createSelector(
  selectedThreadSelectors.getIsNetworkChartEmptyInFullRange,
  selectedThreadSelectors.getJsTracerTable,
  (isNetworkChartEmpty, jsTracerTable) => {
    let visibleTabs = tabSlugs;
    if (isNetworkChartEmpty) {
      visibleTabs = visibleTabs.filter(tabSlug => tabSlug !== 'network-chart');
    }
    if (!jsTracerTable) {
      visibleTabs = visibleTabs.filter(tabSlug => tabSlug !== 'js-tracer');
    }
    return visibleTabs;
  }
);
