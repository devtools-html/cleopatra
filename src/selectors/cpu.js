/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import { createSelector } from 'reselect';

import {
  getThreads,
  getProfileInterval,
  getSampleUnits,
  getMeta,
} from './profile';
import { getThreadSelectors } from './per-thread';
import { computeMaxThreadCPUDelta } from 'firefox-profiler/profile-logic/cpu';

import type { Selector, State, Thread } from 'firefox-profiler/types';

export const getIsCPUUtilizationProvided: Selector<boolean> = createSelector(
  getSampleUnits,
  getMeta,
  getCPUProcessedThreads,
  (sampleUnits, meta, threads) => {
    return (
      sampleUnits !== undefined &&
      // Currently checking the features array for 'cpu' feature should be enough,
      // but in the future we may remove that feature and enable it permanently.
      // Therefore we are also checking the samples table to see if we have CPU
      // delta values.
      ((meta.configuration && meta.configuration.features.includes('cpu')) ||
        threads.some((thread) => thread.samples.threadCPUDelta !== undefined))
    );
  }
);

/**
 * This function returns the list of all threads after the CPU values have been
 * processed. This uses a selector from the per-thread selectors. Because we'll
 * use this selector for every thread, and also need the full state for this call,
 * we can't use the simple memoization from `createSelector`, and instead we
 * need to implement our own simple memoization.
 */
let _threads = null;
let _cpuProcessedThreads = null;
function getCPUProcessedThreads(state: State): Thread[] {
  const threads = getThreads(state);

  if (_threads !== threads || _cpuProcessedThreads === null) {
    // Storing the threads makes it possible to invalidate the memoized value at
    // the right moment.
    _threads = threads;
    _cpuProcessedThreads = threads.map((thread, threadIndex) =>
      getThreadSelectors(threadIndex).getCPUProcessedThread(state)
    );
  }
  return _cpuProcessedThreads;
}

export const getMaxThreadCPUDelta: Selector<number> = createSelector(
  getCPUProcessedThreads,
  getProfileInterval,
  computeMaxThreadCPUDelta
);
