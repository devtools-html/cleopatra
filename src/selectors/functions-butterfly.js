/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import { createSelector } from 'reselect';
import { getDefaultCategory } from './profile';
import { selectedThreadSelectors } from './per-thread';
import { getCallNodeInfo } from '../profile-logic/profile-data';

import type {
  IndexIntoFuncTable,
  Thread,
  IndexIntoCategoryList,
} from '../types/profile';
import type { CallNodeInfo } from '../types/profile-derived';
import type { State } from '../types/state';

const EMPTY_ARRAY = [];

const getImplementationFilteredCallNodeInfo = createSelector(
  selectedThreadSelectors.getImplementationFilteredThread,
  getDefaultCategory,
  (
    { stackTable, frameTable, funcTable }: Thread,
    defaultCategory: IndexIntoCategoryList
  ): CallNodeInfo => {
    return getCallNodeInfo(stackTable, frameTable, funcTable, defaultCategory);
  }
);

const getCallingInformationForFunction = createSelector(
  selectedThreadSelectors.getImplementationFilteredThread,
  getImplementationFilteredCallNodeInfo,
  (_, funcIndex: IndexIntoFuncTable | null) => funcIndex,
  (
    thread,
    { callNodeTable },
    funcIndex
  ): {| callers: IndexIntoFuncTable[], callees: IndexIntoFuncTable[] |} => {
    if (funcIndex === null) {
      return { callers: EMPTY_ARRAY, callees: EMPTY_ARRAY };
    }

    const callers = new Set();
    const callees = new Set();

    for (let i = 0; i < callNodeTable.length; i++) {
      const thisFuncIndex = callNodeTable.func[i];
      const callerNodeIndex = callNodeTable.prefix[i];

      if (callerNodeIndex >= 0) {
        const callerFuncIndex = callNodeTable.func[callerNodeIndex];

        if (thisFuncIndex === funcIndex) {
          callers.add(callerFuncIndex);
        }

        if (callerFuncIndex === funcIndex) {
          callees.add(thisFuncIndex);
        }
      }
    }

    return { callers: [...callers], callees: [...callees] };
  }
);

export const getCallersForFunction = (
  state: State,
  funcIndex: IndexIntoFuncTable | null
) => getCallingInformationForFunction(state, funcIndex).callers;
export const getCalleesForFunction = (
  state: State,
  funcIndex: IndexIntoFuncTable | null
) => getCallingInformationForFunction(state, funcIndex).callees;
