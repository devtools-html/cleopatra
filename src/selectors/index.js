/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import { selectedThreadSelectors as selectedThread } from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';

// Normally we don't need that type because flow can infer types, but on that
// object, it gives weird flow errors that say missing type annotation for X.
// Only adding a type like that seems to solve it.
type Selectors = {
  app: typeof app,
  profile: typeof profile,
  urlState: typeof urlState,
  icons: typeof icons,
  publish: typeof publish,
  zippedProfiles: typeof zippedProfiles,
  selectedThread: typeof selectedThread,
};

export default ({
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
}: Selectors);
