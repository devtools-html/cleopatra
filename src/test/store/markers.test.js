/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow
import { storeWithProfile } from '../fixtures/stores';
import { selectedThreadSelectors } from '../../selectors/per-thread';
import { getProfileWithMarkers } from '../fixtures/profiles/make-profile';

describe('selectors/getMarkerChartTiming', function() {
  function getMarkerChartTiming(testMarkers) {
    const profile = getProfileWithMarkers(testMarkers);
    const { getState } = storeWithProfile(profile);
    return selectedThreadSelectors.getMarkerChartTiming(getState());
  }

  it('has no marker timing if no markers are present', function() {
    expect(getMarkerChartTiming([])).toEqual([]);
  });

  describe('markers of the same name', function() {
    it('puts markers of the same time in two rows', function() {
      // The timing should look like this:
      // 'Marker Name': *------*
      //              : *------*
      const markerTiming = getMarkerChartTiming([
        ['Marker Name', 0, { startTime: 0, endTime: 10 }],
        ['Marker Name', 0, { startTime: 0, endTime: 10 }],
      ]);
      expect(markerTiming).toHaveLength(2);
    });

    it('puts markers of disjoint times in one row', function() {
      // The timing should look like this:
      // 'Marker Name': *------*  *------*
      const markerTiming = getMarkerChartTiming([
        ['Marker Name', 0, { startTime: 0, endTime: 10 }],
        ['Marker Name', 0, { startTime: 15, endTime: 25 }],
      ]);
      expect(markerTiming).toHaveLength(1);
    });

    it('puts markers of overlapping times in two rows', function() {
      // The timing should look like this:
      // 'Marker Name': *------*
      //              :     *------*
      const markerTiming = getMarkerChartTiming([
        ['Marker Name', 0, { startTime: 0, endTime: 10 }],
        ['Marker Name', 0, { startTime: 5, endTime: 15 }],
      ]);
      expect(markerTiming).toHaveLength(2);
    });

    it('puts markers of inclusive overlapping times in two rows', function() {
      // The timing should look like this:
      // 'Marker Name': *--------*
      //              :   *---*
      const markerTiming = getMarkerChartTiming([
        ['Marker Name', 0, { startTime: 0, endTime: 20 }],
        ['Marker Name', 0, { startTime: 5, endTime: 15 }],
      ]);
      expect(markerTiming).toHaveLength(2);
    });
  });

  describe('markers of the different names', function() {
    it('puts them in different rows', function() {
      // The timing should look like this:
      // 'Marker Name A': *------*
      // 'Marker Name B':           *------*
      const markerTiming = getMarkerChartTiming([
        ['Marker Name A', 0, { startTime: 0, endTime: 10 }],
        ['Marker Name B', 0, { startTime: 20, endTime: 30 }],
      ]);
      expect(markerTiming).toHaveLength(2);
      expect(markerTiming[0].name).toBe('Marker Name A');
      expect(markerTiming[1].name).toBe('Marker Name B');
    });
  });

  describe('markers that are crossing the profile start or end', function() {
    it('renders properly markers starting before profile start', function() {
      const markerTiming = getMarkerChartTiming([
        [
          'Rasterize',
          1,
          { category: 'Paint', interval: 'end', type: 'tracing' },
        ],
      ]);
      expect(markerTiming).toEqual([
        {
          name: 'Rasterize',
          // First sample is captured at time 1, so this incomplete
          // marker will start at that same point.
          start: [1],
          end: [1],
          index: [0],
          label: [''],
          length: 1,
        },
      ]);
    });

    it('renders properly markers ending after profile end', function() {
      const markerTiming = getMarkerChartTiming([
        [
          'Rasterize',
          20,
          { category: 'Paint', interval: 'start', type: 'tracing' },
        ],
      ]);
      expect(markerTiming).toEqual([
        {
          name: 'Rasterize',
          start: [20],
          end: [Infinity],
          index: [0],
          label: [''],
          length: 1,
        },
      ]);
    });
  });
});

describe('getProcessedMarkersThread', function() {
  function getProcessedMarkers(testMarkers) {
    const profile = getProfileWithMarkers(testMarkers);
    const { getState } = storeWithProfile(profile);
    return selectedThreadSelectors.getProcessedMarkersTable(getState());
  }

  it('can process Invalidation markers', function() {
    const markers = getProcessedMarkers([
      ['Invalidate http://mozilla.com/script.js:1234', 10, null],
      ['Invalidate self-hosted:2345', 20, null],
      ['Invalidate resource://foo -> resource://bar:3456', 30, null],
    ]);
    expect(markers.time).toEqual([10, 20, 30]);
    expect(markers.data).toEqual([
      {
        type: 'Invalidation',
        url: 'http://mozilla.com/script.js',
        line: '1234',
        startTime: 10,
        endTime: 10,
      },
      {
        type: 'Invalidation',
        url: 'self-hosted',
        line: '2345',
        startTime: 20,
        endTime: 20,
      },
      {
        type: 'Invalidation',
        url: 'resource://foo -> resource://bar',
        line: '3456',
        startTime: 30,
        endTime: 30,
      },
    ]);
  });

  it('can process Bailout markers', function() {
    const markers = getProcessedMarkers([
      [
        'Bailout_ShapeGuard after getelem on line 3666 of resource://foo.js -> resource://bar.js:3662',
        10,
        null,
      ],
      [
        'Bailout_TypeBarrierV at jumptarget on line 1021 of self-hosted:970',
        20,
        null,
      ],
    ]);
    expect(markers.time).toEqual([10, 20]);
    expect(markers.data).toEqual([
      {
        type: 'Bailout',
        bailoutType: 'ShapeGuard',
        where: 'after getelem',
        script: 'resource://foo.js -> resource://bar.js',
        bailoutLine: 3666,
        functionLine: 3662,
        startTime: 10,
        endTime: 10,
      },
      {
        type: 'Bailout',
        bailoutType: 'TypeBarrierV',
        where: 'at jumptarget',
        script: 'self-hosted',
        bailoutLine: 1021,
        functionLine: 970,
        startTime: 20,
        endTime: 20,
      },
    ]);
  });
});
