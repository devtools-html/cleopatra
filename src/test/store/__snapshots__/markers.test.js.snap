// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTimelineVerticalMarkers gets the appropriate markers 1`] = `
Array [
  Object {
    "category": 0,
    "data": Object {
      "category": "Navigation",
      "eventType": "load",
      "innerWindowID": 1,
      "interval": "start",
      "type": "tracing",
    },
    "dur": 1,
    "name": "Load",
    "start": 4,
    "title": null,
  },
  Object {
    "category": 0,
    "data": null,
    "dur": 0,
    "name": "TTI",
    "start": 6,
    "title": null,
  },
  Object {
    "category": 0,
    "data": Object {
      "category": "Navigation",
      "innerWindowID": 1,
      "interval": "start",
      "type": "tracing",
    },
    "dur": 1,
    "name": "DOMContentLoaded",
    "start": 6,
    "title": null,
  },
  Object {
    "category": 0,
    "data": null,
    "dur": 0,
    "name": "Navigation::Start",
    "start": 7,
    "title": null,
  },
  Object {
    "category": 0,
    "data": null,
    "dur": 0,
    "name": "Contentful paint at something",
    "start": 8,
    "title": null,
  },
]
`;
